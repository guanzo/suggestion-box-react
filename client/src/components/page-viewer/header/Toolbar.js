import React, { Component } from 'react';
import { connect } from 'react-redux'
import { Transition } from 'react-transition-group'
import { isAdminSelector } from '@store/user'
import { sortSuggestions } from '@store/suggestions'
import { changeCurrentListType } from '@store/suggestions-admin'
const { 
	SORT_VOTES, SORT_NEW, LIST_APPROVED, LIST_PENDING
} = require('@shared/suggestion-util')

export class Toolbar extends Component {
	constructor(props){
		super(props)
		this.state = {
			sortBy: SORT_VOTES,
			listType: LIST_APPROVED
		}
		this.onSortByChanged = this.onSortByChanged.bind(this)
		this.onListTypeChanged = this.onListTypeChanged.bind(this)
	}
    render() {
		let { hasSuggestions, listType, currentUser } = this.props
		let showSortBy = hasSuggestions && listType === LIST_APPROVED
		const duration = 250;
        return (
            <div className="toolbar flex is-size-7">
				<Transition in={showSortBy} timeout={duration} unmountOnExit>
					{(state) => this.sortBy(state, duration)}
				</Transition>
				{ currentUser.isAdmin ? this.listType() : null }
            </div>
        );
	}
	sortBy(transitionState, duration){
		const defaultStyle = {
			transition: `${duration}ms`,
			opacity: 0,
		}
		const transitionStyles = {
			entering: { opacity: 0 },
			entered:  { opacity: 1 },
		};
		return (
			<div className="flex-center"
				style={{
					...defaultStyle,
					...transitionStyles[transitionState]
				}}
			>
				<span className="m-r-5">Sort by</span>
				<div className="select is-small select-sort-by">
					<select value={this.state.sortBy} 
							onChange={this.onSortByChanged}
					>
						<option value={SORT_VOTES}>Top</option>
						<option value={SORT_NEW}>New</option>
						{/* <option value={SORT_BROADCASTER_VOTES}>Broadcaster likes</option> */}
					</select>
				</div>
			</div>
		)
	}
	listType(){
		return (
			<div className="flex-center m-l-a">
				<span className="m-r-5">Admin</span>
				<div className="select is-small select-list-type">
					<select value={this.state.listType} 
							onChange={this.onListTypeChanged}
					>
						<option value={LIST_APPROVED}>Approved</option>
						<option value={LIST_PENDING}>Pending</option>
					</select>
				</div>
			</div>
		)
	}
	onListTypeChanged(e){
		let listType = e.target.value
		this.setState({ listType })
		this.props.changeCurrentListType(listType)
	}
	onSortByChanged(e){
		let sortBy = e.target.value
		this.setState({ sortBy })
		this.props.sortSuggestions(sortBy)
	}
}

const mapStateToProps = (state) => {
	let { user, suggestions } = state
	let { currentListType } = suggestions

	return {
		currentUser: {
			...user,
			isAdmin: isAdminSelector(state)
		},
		listType: currentListType,
		hasSuggestions: suggestions[currentListType].data.length > 0,
    }
}
const mapDispatchToProps = (dispatch, ownProps) => {
    return {
		changeCurrentListType: (listType)=>dispatch(changeCurrentListType(listType)),
		sortSuggestions: (sortBy)=>dispatch(sortSuggestions(sortBy))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(Toolbar)
