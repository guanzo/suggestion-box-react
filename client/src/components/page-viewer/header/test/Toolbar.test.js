
import React from 'react';
import PropTypes from 'prop-types';
import { createStore } from 'redux';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import {Toolbar} from '../Toolbar';
const { 
	SORT_VOTES, SORT_NEW, LIST_APPROVED, LIST_PENDING
} = require('@shared/suggestion-util')

function mockProps({currentListType = LIST_APPROVED, isAdmin, hasSuggestions = true}){
	return {
		currentUser: {
			isAdmin
		},
		listType: currentListType,
		hasSuggestions
	}
}

describe('<Toolbar />',()=>{
	it('does not show admin dropdown to non-admins', () => {
		const props = { isAdmin: false }
		const wrapper = shallow(<Toolbar {...mockProps(props)}/>)
		expect(wrapper.find('.select-list-type')).toHaveLength(0);
	})
	it('shows admin dropdown to admins', () => {
		const props = { isAdmin: true }
		const wrapper = shallow(<Toolbar {...mockProps(props)}/>)
		expect(wrapper.find('.select-list-type')).toHaveLength(1);
	})
	it('renders correctly',()=>{
		const props = { isAdmin: true }
		const tree = renderer 
		.create(<Toolbar {...mockProps(props)}/>)
		.toJSON();
	  	expect(tree).toMatchSnapshot();
	})
})