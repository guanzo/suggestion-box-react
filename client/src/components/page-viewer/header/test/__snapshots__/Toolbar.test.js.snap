// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toolbar /> renders correctly 1`] = `
<div
  className="toolbar flex is-size-7"
>
  <div
    className="flex-center"
    style={
      Object {
        "opacity": 1,
        "transition": "250ms",
      }
    }
  >
    <span
      className="m-r-5"
    >
      Sort by
    </span>
    <div
      className="select is-small select-sort-by"
    >
      <select
        onChange={[Function]}
        value="votesLength"
      >
        <option
          value="votesLength"
        >
          Top
        </option>
        <option
          value="createdAt"
        >
          New
        </option>
      </select>
    </div>
  </div>
  <div
    className="flex-center m-l-a"
  >
    <span
      className="m-r-5"
    >
      Admin
    </span>
    <div
      className="select is-small select-list-type"
    >
      <select
        onChange={[Function]}
        value="approved"
      >
        <option
          value="approved"
        >
          Approved
        </option>
        <option
          value="pending"
        >
          Pending
        </option>
      </select>
    </div>
  </div>
</div>
`;
